rs/replicated_state/src/replicated_state.rs

```diff
@@ -968,6 +968,67 @@ impl ReplicatedState {
         (expired_callback_count, errors)
     }
 
+    /// Enforces the best-effort message limit by repeatedly shedding the largest
+    /// best-effort messages of the canisters with the largest best-effort memory
+    /// usage until the memory usage drops below the limit.
+    ///
+    /// Time complexity: `O(n * log(n))`.
+    pub fn enforce_best_effort_message_limit(&mut self, limit: NumBytes) {
+        const ZERO: NumBytes = NumBytes::new(0);
+
+        // Check if we need to do anything at all before constructing a priority queue.
+        let mut memory_usage = self
+            .canister_states
```

Comment:
I suppose we are already iterating over the canister states a couple of times in the round, but just to double check: could this become too expensive?

Reply:
We are iterating over all canisters many, many times: every time we compute memory usage (which is a few times per inner round loop); in the scheduler (again, a few times: updating the schedule, filtering active canisters, charging executed canisters); when we aggregate metrics; when we create input and output iterators; probably a couple more I'm missing.

The fix to all this madness is to separate canisters into "active" and "inactive" (by whatever measure, there are at least a few competing definitions); encapsulate all of this within `ReplicatedState`; and then maintain stats the way `CanisterQueues` / `MessagePool` do, so that querying said stats is constant time; and even when iteration (over the set of active canisters) is required, this is still one or two orders of magnitude faster than what we do now.

For now, for better or worse, I guess we just keep iterating over all canisters until someone is bothered enough to push hard for "active canisters" and encapsulation.